### ISSUE TITLE
Bug of selecting targets when computing logits contribution

### ISSUE THREAD

**Issue by xiaoda99** [TomFrederik/unseal#7]
In this piece of code:
https://github.com/TomFrederik/unseal/blob/main/unseal/hooks/common_hooks.py#L259-L276
`            temp = attn_weights[...,None] * value[:,:,None]
            if len(c_proj.shape) == 2:
                c_proj = einops.rearrange(c_proj, '(head_dim num_heads) out_dim -> head_dim num_heads out_dim', num_heads=attn_output.shape[1])
            c_proj = einops.rearrange(c_proj, 'h n o -> n h o')
            temp = temp[0,:,:-1] # could this be done earlier?
            new_temp = []
            for head in tqdm(range(temp.shape[0])):
                new_temp.append([])
                for i in range(math.ceil(temp.shape[1] / batch_size)):
                    out = temp[head, i*batch_size:(i+1)*batch_size] @ c_proj[head]
                    out = out @ vocab_embedding # compute logits
                    new_temp[-1].append(out)
            
                # center logits
                new_temp[-1] = torch.cat(new_temp[-1])
                new_temp[-1] -= torch.mean(new_temp[-1], dim=-1, keepdim=True)
                # select targets
                new_temp[-1] = new_temp[-1][...,torch.arange(len(target_ids)), target_ids]#.to('cpu')`

I found the last line of select targets problematic. new_temp[-1] should be of size (dst_positions, src_positions, vocab_size).  We need to compute the contribution of information movement src -> dst to the logits of target_ids at *dst* positions.  So this line should be:
`new_temp[-1] = new_temp[-1][torch.arange(len(target_ids)), :, target_ids]`
Right?

- Da Xiao

**Comment by xiaoda99** [comment]
I also wrote a cleaner version of this piece of code using einsum:

temp = torch.einsum('bnij,bnjd->nijd', attn_weights, value) # b=1
temp = temp[:, :-1]
new_temp = torch.einsum('nijd,nde->nije', temp, c_proj)
new_temp = new_temp @ vocab_embedding  # nije,ev->nijv
new_temp -= torch.mean(new_temp, dim=-1, keepdim=True)

new_temp = new_temp[:, torch.arange(len(target_ids)), :, target_ids] # our
new_temp = new_temp[:, :, torch.arange(len(target_ids)), target_ids] # your
new_temp = rearrange(new_temp, 'i n j -> n i j')

**Comment by TomFrederik** [comment]
Agreed for the first one, disagree on the second one because of batching issues (see emails I send to you).

Thanks a lot for catching this one! :))
